src/module_1/module_1_meteo_api.py

```
@@ -1,7 +1,176 @@
-""" This is a dummy example """
+import requests
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib.dates as mdates
+import matplotlib.colors as mcolors
+
+
+VARIABLES = ["temperature_2m_mean", "precipitation_sum", "soil_moisture_0_to_10cm_mean"]
+MODELS = [
+    "CMCC_CM2_VHR4",
+    "FGOALS_f3_H",
+    "HiRAM_SIT_HR",
+    "MRI_AGCM3_2_S",
+    "EC_Earth3P_HR",
+    "MPI_ESM1_2_XR",
+    "NICAM16_8S",
+]
+
+years = pd.date_range(start="1950-01-01", end="2050-12-31")
+
+API_URL = "https://climate-api.open-meteo.com/v1/climate"
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+
+
+def get_data_meteo_api(city: str) -> None:
+    df = model_data(city)
+    df["date"] = pd.to_datetime(df["date"])
+    df = df.resample("M", on="date").mean()
+    plot_climate_data(df, city)
+    plt.show()
+
+
+def call_api(city: str, model: str) -> dict:
+    parameters = {
+        "latitude": COORDINATES[city]["latitude"],
+        "longitude": COORDINATES[city]["longitude"],
+        "start_date": "1950-01-01",
+        "end_date": "2050-12-31",
+        "model": model,
+        "timezone": "Europe/Berlin",
+        "daily": VARIABLES,
+    }
+
+    response = requests.get(url=API_URL, params=parameters)
+
+    # Check if the status code is 200
+    if response.status_code != 200:
+        if response.status_code == 429:
+            print("API rate limit exceeded.")
+            raise Exception("API rate limit exceeded")
```

it would be a good idea to implement a retry logic

---

I am not sure I'd raise an exception